ramips: mt7621: add support for Netgear R6800

CHKDSK88 opened the pull request:
This PR adds support for the Netgear R6800, aka Netgear AC1900 and
R6800-100PES.

Specification:
- SoC: MediaTek MT7621AT (880 MHz)
- Falsh: 128 MiB NAND
- RAM: 256 MiB
- Wireless: MediaTek MT7615EN b/g/n , MediaTek MT7615EN an+ac
- LAN speed: 10/100/1000
- LAN ports: 4
- WAN speed: 10/100/1000
- WAN ports: 1
- USB 2.0
- USB 3.0
- Serial baud rate of Bootloader and factory firmware: 57600

Installation:
- apply factory image via stock web-gui.

Back to stock:
- nmrpflash can be used to recover to the netgear firmware if a broken
image was flashed.

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 committed:
ramips: mt7621: add support for Netgear R6800

This patch adds support for the Netgear R6800, aka Netgear AC1900 and
R6800-100PES.

Specification:
- SoC: MediaTek MT7621AT (880 MHz)
- Flash: 128 MiB NAND
- RAM: 256 MiB
- Wireless: MediaTek MT7615EN b/g/n , MediaTek MT7615EN an+ac
- LAN speed: 10/100/1000
- LAN ports: 4
- WAN speed: 10/100/1000
- WAN ports: 1
- USB 2.0
- USB 3.0
- Serial baud rate of Bootloader and factory firmware: 57600

Known issues:
- Device has 3 wifi LEDs: Wifi 5Ghz, Wifi 2.4Ghz and Wifi on/off.
  Wifi on/off is not used.

Installation:
- apply factory image via stock web-gui.

Back to stock:
- nmrpflash can be used to recover to the stock Netgear firmware.

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

adrianschmutzler commented on target/linux/ramips/dts/mt7621_netgear_r6800.dts:
```diff
@@ -0,0 +1,81 @@
+
+&led_wps {
+	label = "r6800:white:wps";
+};
```
Since there is only one line per LED left in the dtsi and we need all those node labels, I would strongly prefer having the whole LED block in the DTS file.

CHKDSK88 replied:
Will be done.

adrianschmutzler replied:
BTW: How probable is it that we will receive other devices using the DTSI soon? I just wonder whether it wouldn't be easier to put everything into DTS and create DTSI when we have a second device that needs it.

CHKDSK88 replied:
[Techinfodepot](http://en.techinfodepot.shoutwiki.com/wiki/Netgear_R6800) says, that the same board is used in 5 more devices. But I don't have other boards for tests. 

It is possible to add support for all but I'm not sure when it will happen.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_netgear_sercomm_bzv.dtsi:
```diff
@@ -0,0 +1,230 @@
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x4>;
```
nothing at 0xe000 or 0x28?

CHKDSK88 replied:
Nothing.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_netgear_sercomm_bzv.dtsi:
```diff
@@ -0,0 +1,230 @@
+			ralink,function = "gpio";
+		};
+	};
+};
```
state_default is already defined in mt7621.dtsi, so you use it here:

```
&state_default {
	gpio {
		ralink,group = "uart3", "uart2", "jtag", "wdt", "spi";
		ralink,function = "gpio";
	};
};
```

Haven't found why you need "spi" here?

CHKDSK88 replied:
Sorry. I forgot remove it.

adrianschmutzler replied:
Still need to change that to:
```
&state_default {
	gpio {
		ralink,group = "uart3", "uart2", "jtag", "wdt";
		ralink,function = "gpio";
	};
};
```

adrianschmutzler commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -525,6 +525,18 @@ define Device/netgear_r6350
+define Device/netgear_r6800
+  $(Device/netgear_sercomm_nand)
+  DEVICE_MODEL := R6800
+  SERCOMM_HWNAME := R6950
```
Is the deviation from device name by intention?

CHKDSK88 replied:
This came from OFW header:

````binwalk -e R6800-V1.2.0.62_1.0.1.img

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             Sercomm firmware signature, version control: 256, download control: 0, hardware ID: "BZV", hardware version: 0x4100, firmware version: 0x62, starting code segment: 0x0, code size: 0x7300
63            0x3F            Sercomm firmware signature, version control: 0, download control: 0, hardware ID: "", hardware version: 0x0, firmware version: 0x0, starting code segment: 0x0, code size: 0x0
512           0x200           Zip archive data, at least v2.0 to extract, compressed size: 32714253, uncompressed size: 83886080, name: R6950.bin
32714893      0x1F3308D       End of Zip archive

````

adrianschmutzler replied:
Okay.

adrianschmutzler commented on target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:
```diff
@@ -52,6 +52,17 @@ netgear,r6850|\
+	ucidef_set_led_netdev "wlan2g" "WiFi 2.4GHz" "$boardname:white:wlan2g" "wlan0"
+	ucidef_set_led_netdev "wlan5g" "WiFi 5GHz" "$boardname:white:wlan5g" "wlan1"
+	ucidef_set_led_usbport "usb2" "USB 2.0" "$boardname:white:usb2" "usb1-port2"
+	ucidef_set_led_usbport "usb3" "USB 3.0" "$boardname:white:usb3" "usb1-port1"
```
WiFi and USB leds should be configurable in DTS instead.

CHKDSK88 replied:
Done.

adrianschmutzler commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -229,6 +230,7 @@ ramips_setup_macs()
 	netgear,r6350|\
+	netgear,r6800|\
 	netgear,r6850)
 		wan_mac=$(macaddr_add "$(mtd_get_mac_binary factory 0x4)" 2)
```
Nothing in 0xe006 or 0x2e?

CHKDSK88 replied:
Nope.

CHKDSK88 commented:
@adrianschmutzler 

All changes have been done.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_netgear_r6800.dts:
```diff
@@ -0,0 +1,118 @@
+			label = "r6800:white:power";
+		};
+
+		led_wan_orange: wan_orange {
```
Okay, so now you can remove all node labels you do not use, i.e. all except power...

adrianschmutzler commented on target/linux/ramips/dts/mt7621_netgear_r6800.dts:
```diff
@@ -0,0 +1,118 @@
+			label = "r6800:white:wifi";
+		};
+	};
+
```
Remove empty line here.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_netgear_sercomm_bzv.dtsi:
```diff
@@ -0,0 +1,157 @@
+		reg = <1>;
+		#trigger-source-cells = <0>;
+	};
+	hub_port2: port@2 {
```
Add empty line before hub_port2.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_netgear_sercomm_bzv.dtsi:
```diff
@@ -0,0 +1,157 @@
+			bias-pull-up;
+		};
+	};
+
```
Remove empty line.

CHKDSK88 commented:
Done. Sorry for extra "spi".

adrianschmutzler commented:
And now remove the enclosing pinctrl please, as explained in the commit (use &state_default directly). :-)

adrianschmutzler commented on target/linux/ramips/dts/mt7621_netgear_r6800.dts:
```diff
@@ -0,0 +1,117 @@
+			label = "r6800:white:wan";
+		};
+
+		wifi2g {
```
If the label property uses "wlan2g", I'd be consistent with the node name. I do not have a preference for wlan2g or wifi2g, but it should be consistent. Same for 5g below.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_netgear_r6800.dts:
```diff
@@ -0,0 +1,117 @@
+			label = "r6800:white:wps";
+		};
+
+		wifi {
```
What's the purpose of this LED in addition to the two WiFi leds above? Also consider that for proper choice of names ...

CHKDSK88 replied:
Router have 3 wifi LEDs: Wifi 5Ghz, Wifi 2,4Ghz and Wifi on/off LED.

adrianschmutzler replied:
Hmm, okay, then maybe just keep "wifi" or "wlan" for this one.
Maybe add the comment to commit message:
"Device has 3 wifi LEDs: Wifi 5Ghz, Wifi 2.4Ghz and Wifi on/off"

adrianschmutzler commented:
Sorry, found something again.

CHKDSK88 commented:
> Sorry, found something again.

Sorry for so many little mistakes. I had couple months break.

vgmz95 commented:
Hello, I have a Netgear R6700v2 (it uses the same board) and I installed this version of openwrt on it. 
It runs fine and  I haven't found any issues so far, aside from the WiFi LED issue.
If you need me to test something else I'll gladly help.

CHKDSK88 commented:
@vgmz95 
Thank You for testing!

According [Techinfodepot](http://en.techinfodepot.shoutwiki.com/wiki/Netgear_R6700_v2), the only differences are unpopulated USB 2.0 connector and blank place for USB 2.0 LED. Is that correct?

adrianschmutzler commented:
@vgmz95 @CHKDSK88 There is an unfinished, abandoned PR for R6700 v2 here:
https://github.com/openwrt/openwrt/pull/2545

If the PR for R6800 is merged, I might revive the other one and finish it myself if @vgmz95 is willing to test.

I will have a look at what's similar and different for those, too, maybe we can benefit from that for setting up DTS(I).

adrianschmutzler commented:
Commit message: Falsh -> Flash

adrianschmutzler commented:
@CHKDSK88 Having had a look at the R6700 v2 PR and your patches, I think it makes sense to keep DTSI and DTS as you have proposed them.

Since R6700v2 has no USB, you should move the xhci node from DTSI to DTS.

Despite, please move the state_default node to DTS, too, as it's dependent on the LEDs there. After you've done that, I will put together a patch for the R6700 v2 based on it.

adschm committed:
ramips: add support for NETGEAR R6700v2/AC2400

SoC: MediaTek MT7621AT
RAM: 256M DDR3
FLASH: 128M NAND
WiFi: MediaTek MT7615N an+ac
MediaTek MT7615N bgn
ETH: MediaTek MT7621AT
BTN: 1x Connect (WPS), 1x WLAN, 1x Reset
LED: Power (white/amber), WAN(white/amber), 2.4G(white), 5G(white),
USB(white) , GuestWifi(white) 4x LAN(white/amber), Wifi Button(white),
WPS Button(white)

Installation:

Login to netgear webinterface and flash factory.img

Based on a discontinued GitHub Pull Request by
kuyokushin <codenamezero@protonmail.com>

https://github.com/openwrt/openwrt/pull/2545

NOTE: Netgear R6700 v2 have five clones: R6900 v2, R7450, Nighthawk
AC2400, Nighthawk AC2100 and already added R6800. Rest of them  should
be really easy supportable. Image for R6700v2 should work perfectly with
them. Please refer:

https://github.com/openwrt/openwrt/pull/2614

Tested-by: Víctor Gibrán <victorgibranmz@hotmail.com> [R6700v2]
Tested-by: John Landrum <jl31m10@yahoo.com> [AC2400]
Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>
[add guest led to mt7621_netgear_r6700-v2.dts end edit commit message]
Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

vgmz95 commented:
@CHKDSK88 That's correct, R6700v2 has no USB2.0 at the back and no USB2.0 LED.

@adrianschmutzler I'm willing to test that in my device when it's ready.

I also found some issues regarding the 2.4Ghz radio and I'll report them in the mt76 repo.

adrianschmutzler commented:
@vgmz95 I've put a first attempt into one of my staging trees here:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/r6700v2

If you like, you can test that one already, particularly with respect to the leds and buttons.

vgmz95 commented:
@adrianschmutzler 
I've installed and tested that version and here's what I found: 

- Working LEDs: ON, WAN, WiFi 2GHz, WiFi 5GHz, USB3.0, LAN1-4, WLAN, WPS
- Missing LED: Guest WiFi (no big deal I guess)
- Working buttons: WiFi, WPS , reset

Other issues: 
- Can't read syslog 
- 2.4GHz issues

CHKDSK88 commented:
@adrianschmutzler 
@vgmz95 

I made some changes. Usb hub isn't required because is in mt7621.dtsi, so I removed it.

>Despite, please move the state_default node to DTS, too, as it's dependent on the LEDs there.

IMO all devices will have the same state_default node.

CHKDSK88 commented:
One more thing:

@vgmz95
Could You paste result of "hexdump /dev/mtd6" here?

adrianschmutzler commented:
> > Despite, please move the state_default node to DTS, too, as it's dependent on the LEDs there.

> IMO all devices will have the same state_default node.

I do not doubt that, but since the LED definitions with most of the gpio properties are located in the DTS files and the state_default block is not huge, it is more logical to me to put it there. But I won't force you if you feel otherwise.

adrianschmutzler commented:
> I made some changes. Usb hub isn't required because is in mt7621.dtsi, so I removed it.

Nice, good find.

One other thing:
In the PR for r6700v2, there is a power led (green) set up with gpio0 22. Based on the test reports here, this is wrong?

CHKDSK88 commented:
@adrianschmutzler 

>I do not doubt that, but since the LED definitions with most of the gpio properties are located in the DTS files and the state_default block is not huge, it is more logical to me to put it there. 

I changed. You are the core developer. ;) 

>In the PR for r6700v2, there is a power led (green) set up with gpio0 22. Based on the test reports here, this is wrong?

IMO green led is mistake. Please look #2545 PR message: "Currently Power,WAN,USB, leds cannot be controlled via gpio as they are connected through some gpio expander."

One more thing:

```
binwalk R6700v2-V1.2.0.62_1.0.1/R6700v2-V1.2.0.62_1.0.1.img 

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             Sercomm firmware signature, version control: 256, download control: 0, hardware ID: "BZV", hardware version: 0x4100, firmware version: 0x62, starting code segment: 0x0, code size: 0x7300
512           0x200           Zip archive data, at least v2.0 to extract, compressed size: 32714253, uncompressed size: 83886080, name: R6950.bin
32714893      0x1F3308D       End of Zip archive, footer length: 22
```

I decided to change SERCOMM_HWNAME to R6950. This is value from stock header.

CHKDSK88 commented:
I make some investigation: 

```

binwalk _R6900v2-V1.2.0.62_1.0.1.zip.extracted/R6900v2-V1.2.0.62_1.0.1.img 

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             Sercomm firmware signature, version control: 256, download control: 0, hardware ID: "BZV", hardware version: 0x4100, firmware version: 0x62, starting code segment: 0x0, code size: 0x7300
512           0x200           Zip archive data, at least v2.0 to extract, compressed size: 32714253, uncompressed size: 83886080, name: R6950.bin
32714893      0x1F3308D       End of Zip archive, footer length: 22
```
```
binwalk _R7450-V1.2.0.62_1.0.1.zip.extracted/R7450-V1.2.0.62_1.0.1.img 

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             Sercomm firmware signature, version control: 256, download control: 0, hardware ID: "BZV", hardware version: 0x4100, firmware version: 0x62, starting code segment: 0x0, code size: 0x7300
512           0x200           Zip archive data, at least v2.0 to extract, compressed size: 32714253, uncompressed size: 83886080, name: R6950.bin
32714893      0x1F3308D       End of Zip archive, footer length: 22
```

Base on Netgear manuals and OFW images for R6700v2, R6900v2 and R7450: This three devices are clones, I think. Image for R6700v2 shout perfectly fit to R6900v2 and R7450. But at this moment we don't have testers for this devices.

vgmz95 commented:
@CHKDSK88  
Here's the hexdump output 
[mtd6.txt](https://github.com/openwrt/openwrt/files/4007287/mtd6.txt)

And as far as I know, the R6700v2. R6800 and R6900v2 are clones and share the same stock img

achterin commented:
@CHKDSK88 are the md5sums of the vendor images identical?

CHKDSK88 commented:
@achterin 

All four routers use the same img:
````
5529fe9350846c42dcbd1072f98311e2  R6700v2-V1.2.0.62_1.0.1/R6700v2-V1.2.0.62_1.0.1.img
5529fe9350846c42dcbd1072f98311e2  R6800-V1.2.0.62_1.0.1/R6800-V1.2.0.62_1.0.1.img
5529fe9350846c42dcbd1072f98311e2  _R6900v2-V1.2.0.62_1.0.1.zip.extracted/R6900v2-V1.2.0.62_1.0.1.img
5529fe9350846c42dcbd1072f98311e2  _R7450-V1.2.0.62_1.0.1.zip.extracted/R7450-V1.2.0.62_1.0.1.img
````

@vgmz95 
Thank You for dump. I consider using mtd concat but last partition isn't unused.

Could You confirm, that now work everything?

achterin commented:
@CHKDSK88 then its pretty much confirmed that all devices can be supported with one openwrt image. it was the same for R6260, R6350 & R6850, all devices used the same sercomm hardware id, are board identical and share the same fccid. this is also true for the four devices you posted. imho it should be safe to add device definitions for R6900v2 & R7450.
@adrianschmutzler your opinion on that matter?

adrianschmutzler commented:
> IMO green led is mistake. Please look #2545 PR message: "Currently Power,WAN,USB, leds cannot be controlled via gpio as they are connected through some gpio expander."

Okay. I read that, too, and interpreted it the same way. Just wanted to discuss it here before forgetting about that. :-)

> I decided to change SERCOMM_HWNAME to R6950. This is value from stock header.

Fine. The value from R6700v2 PR supposedly has only been copy/pasted from the device name anyway.

> then its pretty much confirmed that all devices can be supported with one openwrt image. it was the same for R6260, R6350 & R6850, all devices used the same sercomm hardware id, are board identical and share the same fccid. this is also true for the four devices you posted. imho it should be safe to add device definitions for R6900v2 & R7450.

Although I also believe that you are right, I won't merge device support without on-device test.

Anyway, I would focus on getting R6800 and R6700 v2 merged for now. It might be helpful to include your findings about the clones into the commit message of R6700 v2, so it is preserved for the future.

I wonder whether we can already use the information about the clones to improve DTS(I) files. Since most of the definitions in r6700-v2.dts are leds, though, I do not really see a good way to do that at the moment.

adrianschmutzler commented on package/kernel/linux/modules/other.mk:
```diff
@@ -1230,3 +1230,18 @@ define KernelPackage/it87-wdt/description
+endef
+
+define KernelPackage/pinctrl-sx150x/description
+ This driver adds support for Semtech SX150x-series I2C GPIO expanders.
```
Nitpick: Double space indent required here.

vgmz95 commented:
@CHKDSK88 I can confirm that all LEDs and buttons are working correctly :)

CHKDSK88 commented:
@adrianschmutzler 

I hope that PR is ready for merge now. :)

kuyokushin commented:
Sorry for the late reply, I was unavailable for a few weeks. I can confirm that your commit for the r6700-v2 is correct and flash tested. But as far as I can see the triggers for wlan2g/wlan5g e.g. linux,default-trigger = "phy1tpt"; are not working as phy0tpt/phy1tpt are not available in current driver.

adrianschmutzler commented:
@blocktrron Can you review the GPIO expander and the corresponding kernel patch here? The rest looks fine to me.

adrianschmutzler commented:
@CHKDSK88 Are the triggers working for your R6800?

CHKDSK88 commented:
@adrianschmutzler 
At this moment wlan led trigger isn't working. I was pretty sure that I tested it. But at this moment I can't reproduce it.

@vgmz95 
Are wlan led triggers work in Your r6700v2?

kuyokushin commented:
Triggers are working, e.g. using phy0tx/phy1tx, but I don't see phy0tpt/phy1tpt as an option given by the driver.

adrianschmutzler commented:
Seems like phyXtpt is used on some device with mt7621, but not too many, so that's not too helpful:
```
adsc@buildfff:/data/openwrt$ grep -rn "phy.radio" target/linux/ramips/dts/mt7621*
target/linux/ramips/dts/mt7621_asus_rt-acx5p.dtsi:52:                   linux,default-trigger = "phy0radio";
target/linux/ramips/dts/mt7621_asus_rt-acx5p.dtsi:58:                   linux,default-trigger = "phy1radio";
target/linux/ramips/dts/mt7621_elecom_wrc-1167ghbk2-s.dts:29:                   linux,default-trigger = "phy0radio";
target/linux/ramips/dts/mt7621_iodata_wnpr2600g.dts:41:                 linux,default-trigger = "phy0radio";
target/linux/ramips/dts/mt7621_iodata_wnpr2600g.dts:47:                 linux,default-trigger = "phy1radio";
target/linux/ramips/dts/mt7621_iptime_a6ns-m.dts:38:                    linux,default-trigger = "phy0radio";
target/linux/ramips/dts/mt7621_iptime_a6ns-m.dts:44:                    linux,default-trigger = "phy1radio";
adsc@buildfff:/data/openwrt$ grep -rn "phy.tpt" target/linux/ramips/dts/mt7621*
target/linux/ramips/dts/mt7621_netgear_ex6150.dts:55:                   linux,default-trigger = "phy1tpt";
target/linux/ramips/dts/mt7621_netgear_ex6150.dts:66:                   linux,default-trigger = "phy0tpt";
target/linux/ramips/dts/mt7621_netgear_sercomm_ayx.dtsi:64:                     linux,default-trigger = "phy0tpt";
target/linux/ramips/dts/mt7621_netgear_sercomm_chj.dtsi:42:                     linux,default-trigger = "phy0tpt";
target/linux/ramips/dts/mt7621_xzwifi_creativebox-v1.dts:46:                    linux,default-trigger = "phy0tpt";
target/linux/ramips/dts/mt7621_xzwifi_creativebox-v1.dts:52:                    linux,default-trigger = "phy1tpt";
```

However, if it's not working at the moment I'd recommend to use phyXradio instead for the device support PR. If the situation for tpt trigger changes, we can easily patch it in again ...

adrianschmutzler commented:
Despite, I've removed my r6700v2 branch in the staging repo, as it makes more sense to maintain it here (thanks for taking that up).

vgmz95 commented:
@CHKDSK88 I tried these triggers on the WLAN LED: netdev, phyassoc, phyradio, phyrx/phytx, timer
and all of them seemed to work fine. 
And as @kuyokushin said, theres no phy0tpt/phy1tpt trigger

CHKDSK88 commented:
@adrianschmutzler 

>However, if it's not working at the moment I'd recommend to use phyXradio instead for the device support PR. If the situation for tpt trigger changes, we can easily patch it in again ...

Done.

Another question: 
Is allowed to backport "add support for device" commits to 19.07 branch?

adrianschmutzler commented:
> Is allowed to backport "add support for device" commits to 19.07 branch?

It is "allowed" by meaning it's not forbidden, but generally device support is not backported. The time saved by this can be invested in bringing other features and devices to master.
Thus, in order to have your device backported while all the others aren't, you would have to provide a very good reason for it.

CHKDSK88 commented:
> you would have to provide a very good reason for it.

I don't have any good reason. OpenWRT have 1,5 years periods between stable releases. It's quite long time. Using snapshot releases isn't convenient for regular users. That's only reason. ;)

Yohan54 commented:
Hi,
I just wanted to note that I have tested the netgear-r6800 branch on a Netgear Nighthawk AC2400 and works without any issues that I can see so far. The one thing when trying to clone, I kept getting the openwrt-1 branch which did not include R6800. But when I downloaded the clone via zip it did include R6800.

Basically, I just follow steps on OpenWrt under developer steps. I actually complied the image under Windows 10 WSL, again no issues. I also installed Luci essentials without any major issues.

So, if anyone needs logs or any outputs of commands, let me know.

CHKDSK88 commented:
@Yohan54 

Thank You for testing. 

Do You have [Netgear_Nighthawk_AC2400 from wikidevi](https://wikidevi.wi-cat.ru/Netgear_Nighthawk_AC2400)?

It looks like exactly R6700 clone. 

@adrianschmutzler 
What is current policy about clones?

The approach presented bellow will be ok?
```
define Device/netgear_ac2400
  $(Device/netgear_sercomm_nand)
  DEVICE_MODEL := Nighthawk AC2400
  DEVICE_DTS := mt7621_netgear_r6700-v2
  SERCOMM_HWNAME := R6950
  SERCOMM_HWID := BZV
  SERCOMM_HWVER := A001
  SERCOMM_SWVER := 0x1032
  IMAGE_SIZE := 40960k
  DEVICE_PACKAGES += kmod-mt7615e kmod-pinctrl-sx150x
endef
TARGET_DEVICES += netgear_ac2400
```

adrianschmutzler commented:
> What is current policy about clones?

It depends. If everything is the same (same stock firmware image, some HWNAME etc.) one might just add DEVICE_ALT0_VENDOR and DEVICE_ALT0_MODEL with the new name. This will create an additional entry in "make menuconfig", but won't create a second image.

The alternative would be to create a second device block as you did, which would then also create a second image. In this case, I would insist on doing it consistently, and thus also create a new DTS (with appropriate model and compatible, rest may be derived from a DTSI) and the corresponding base-files entries for the new compatible. This would have the clone as real "separate device", with the disadvantage of additional (=wasted) CPU time for creating the same image again.

Which way of these two is the better one is somehow subjective. I typically decide based on how probable it is that the user will find the correct image if there is only one: If e.g. the v2 is a clone of the v1, it might not be necessary to create a separate image for the v2. If one device is called vendor-AB model-CD, while the other is called vendor-WX model-YZ, it might be better to have separate images. Your case is somewhat in the middle between those.

If you decide for ALT0 approach, you should already include it here. If you decide for a separate image, I'd recommend to wait until this PR is merged to not make it more complicated (and thus delay it merge).

Yohan54 commented:
On Friday, January 17, 2020, 02:56:40 AM EST, Paweł Dembicki <notifications@github.com> wrote:  
 
 
@Yohan54

Thank You for testing.

Do You have Netgear_Nighthawk_AC2400 from wikidevi? Yes

It looks like exactly R6700 clone.

@adrianschmutzler
What is current policy about clones?

The approach presented bellow will be ok?
define Device/netgear_ac2400
  $(Device/netgear_sercomm_nand)
  DEVICE_MODEL := Nighthawk AC2400
  DEVICE_DTS := mt7621_netgear_r6700-v2
  SERCOMM_HWNAME := R6950
  SERCOMM_HWID := BZV
  SERCOMM_HWVER := A001
  SERCOMM_SWVER := 0x1032
  IMAGE_SIZE := 40960k
  DEVICE_PACKAGES += kmod-mt7615e kmod-pinctrl-sx150x
endef
TARGET_DEVICES += netgear_ac2400

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.

Yohan54 commented:
Just let me know, if you need any output information from my device.
My current setup is the OpenWrt router is connected in bridge mode with 5Ghz radio enable only, no need for 2.4 Ghz but can enable for testing purpose. I notice the radio is using the generic MAC80211 driver.
    On Friday, January 17, 2020, 06:32:22 AM EST, Adrian Schmutzler <notifications@github.com> wrote:  
 
 

What is current policy about clones?


It depends. If everything is the same (same stock firmware image, some HWNAME etc.) one might just add DEVICE_ALT0_VENDOR and DEVICE_ALT0_MODEL with the new name. This will create an additional entry in "make menuconfig", but won't create a second image.

The alternative would be to create a second device block as you did, which would then also create a second image. In this case, I would insist on doing it consistently, and thus also create a new DTS (with appropriate model and compatible, rest may be derived from a DTSI) and the corresponding base-files entries for the new compatible. This would have the clone as real "separate device", with the disadvantage of additional (=wasted) CPU time for creating the same image again.

Which way of these two is the better one is somehow subjective. I typically decide based on how probable it is that the user will find the correct image if there is only one: If e.g. the v2 is a clone of the v1, it might not be necessary to create a separate image for the v2. If one device is called vendor-AB model-CD, while the other is called vendor-WX model-YZ, it might be better to have separate images. Your case is somewhat in the middle between those.

If you decide for ALT0 approach, you should already include it here. If you decide for a separate image, I'd recommend to wait until this PR is merged to not make it more complicated (and thus delay it merge).

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.

CHKDSK88 commented:
@adrianschmutzler 

I choose ALT0 solution because this devices are ideal clones. I reworked last commit.

@Yohan54 

Where You find that info? Do You use "iwinfo" tool?

Yohan54 commented:
@CHKDSK88 
yes, here is the output.
root@OpenWrt:~# iwinfo
wlan0     ESSID: unknown
          Access Point: 78:D2:94:85:69:04
          Mode: Client  Channel: unknown (unknown)
          Tx-Power: 6 dBm  Link Quality: unknown/70
          Signal: unknown  Noise: unknown
          Bit Rate: unknown
          Encryption: unknown
          Type: nl80211  HW Mode(s): 802.11bgn
          Hardware: 14C3:7615 7615:14C3 [Generic MAC80211]
          TX power offset: unknown
          Frequency offset: unknown
          Supports VAPs: yes  PHY name: phy0

wlan1     ESSID: "OpenWrt_5G"
          Access Point: 78:D2:94:85:69:05
          Mode: Master  Channel: 36 (5.180 GHz)
          Tx-Power: 20 dBm  Link Quality: 57/70
          Signal: -53 dBm  Noise: -90 dBm
          Bit Rate: 866.7 MBit/s
          Encryption: WPA2 PSK (CCMP)
          Type: nl80211  HW Mode(s): 802.11nac
          Hardware: 14C3:7615 7615:14C3 [Generic MAC80211]
          TX power offset: unknown
          Frequency offset: unknown
          Supports VAPs: yes  PHY name: phy1

Yohan54 commented:
Also, I did observe that is branch is using kernel 4.14.160 when Openwrt snapshot package kernel is 4.14.162

CHKDSK88 commented:
It isn't generic driver. Iwinfo doesn't have the mt7615e in database. I will prepare patch for this.

One more thing:
@vgmz95 @kuyokushin @Yohan54
I would like add You to "tested by" field. Real name and email is required if You decide to do it.

vgmz95 commented:
@CHKDSK88 I'm ok with it 👍, my real name and email address are in my GitHub profile

Yohan54 commented:
> 
> 
> It isn't generic driver. Iwinfo doesn't have the mt7615e in database. I will prepare patch for this.
> 
> One more thing:
> @vgmz95 @kuyokushin @Yohan54
> I would like add You to "tested by" field. Real name and email is required if You decide to do it.

@CHKDSK88 same here 👍, real name and email in my GitHub profile.

Yohan54 commented:
> 
> 
> It isn't generic driver. Iwinfo doesn't have the mt7615e in database. I will prepare patch for this.
> 
> One more thing:
> @vgmz95 @kuyokushin @Yohan54
> I would like add You to "tested by" field. Real name and email is required if You decide to do it.

@CHKDSK88 Got it, definitely using mt7615e driver

lspci | grep -i mediatek
01:00.0 Unclassified device [0002]: MEDIATEK Corp. Device 7615
02:00.0 Unclassified device [0002]: MEDIATEK Corp. Device 7615

modinfo mt7615e
module:         /lib/modules/4.14.160/mt7615e.ko
license:        Dual BSD/GPL
depends:        mac80211,mt76,cfg80211

Aleksandro1 commented:
Hello,

When is the chance for the final version of the R6800 netgear software?

CHKDSK88 commented:
Hi @Aleksandro1 

@adrianschmutzler ask @blocktrron for kernel package review. I think, it might be last blocker.

But I can't tell You when PR will be merged.

adrianschmutzler commented on package/kernel/linux/modules/other.mk:
```diff
@@ -1225,3 +1225,18 @@ define KernelPackage/it87-wdt/description
+define KernelPackage/pinctrl-sx150x
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=Semtech SX150x-series I2C GPIO expanders
+  DEPENDS:= +kmod-i2c-core @!LINUX_4_9
```
You can remove the @!LINUX_4_9 now, as kernel 4.9 has been removed in master.

adrianschmutzler commented:
@CHKDSK88 The commit title for the R6700v2 patch is too long. Either remove the Nighthawk entirely there or try to make it shorter, e.g. remove "mt7621:", use "Netgear R6700 v2/Nighthawk AC2400" or whatever to prevent the title from being wrapped into the commit message.

CHKDSK88 commented:
@adrianschmutzler 

Done.

pepe2k commented on package/kernel/linux/modules/other.mk:
```diff
@@ -1225,3 +1225,18 @@ define KernelPackage/it87-wdt/description
 $(eval $(call KernelPackage,it87-wdt))
+
+
+define KernelPackage/pinctrl-sx150x
```
If this has to be built-in, why not just enable it in subtarget kernel config instead of making it a package?

CHKDSK88 replied:
It could be because it only depends on i2c. I will rework this without this patch.

Yohan54 commented:
Just providing an update of testing on Netgear Nighthawk AC2400

Using the latest R6800 image build of this branch.
Firmware Version: OpenWrt SNAPSHOT unknown / LuCI Master git-20.022.75744-8947526
Kernel Version: 4.14.166

Only issue that I have been seeing is that only channel for 5ghz is channel 36 width 80mhz is stable. From what I have been reading on OpenWrt that it possible to have stable 5ghz on some routers and not on others. I briefly test 2.4ghz, see no issue other than just want strictly use 5ghz.

DEV team, let me know if you need any logs

CHKDSK88 commented:
@adrianschmutzler 

Are patches ready for merge?

adrianschmutzler commented:
@CHKDSK88 I have no objections, but I'm just not qualified to review the gpio-expander, and I do not want to merge it without understanding the code for this node at all.

Yohan54 commented:
Hi,
The mt76 drivers (mt7615e) not stable when device is configured to use channel 36, AC, 80 MHz channel bandwidth.
I am getting the same issue as per here: https://github.com/openwrt/mt76/issues/344 and here: https://github.com/openwrt/mt76/issues/142.
Do you know of mt7615e is supported to use 80 MHz channel bandwidth.

vgmz95 commented:
In my case I'm having a lot of [issues](https://github.com/openwrt/mt76/issues/339)  with the 2GHz radio. I'm using the latest version of the driver.

Regarding the 5GHz radio, in my case it works fine using the channel 36,ac, 80 MHz

Yohan54 commented:
@vgmz95 what netgear router you have?

vgmz95 commented:
> @vgmz95 what netgear router you have?

Netgear R6700v2

Aleksandro1 commented:
Hi everyone, I am only asking if there is a light in the tunnel ;) about the release of the stable version of openwrt for netgear 6800?

Yohan54 commented:
Can only comment of compiling this branch is not difficult to do, instructions are not that complicated and used Linux subsystem of Windows 10. As  for stable version, I do not believe it is close enough for stable release. As for Openwrt snapshot image, I for one don't see a reason why it can't be pushed for downloading.

CHKDSK88 commented:
PR wait for review. I follow @pepe2k instruction, but @adrianschmutzler wait for a core developer, who will review gpio expander part of dts.

I think, this is last blocker.

adrianschmutzler commented:
> but @adrianschmutzler wait for a core developer, who will review gpio expander part of dts.

Doesn't have to be a core developer. If someone else is able to review that part and willing to provide a Reviewed-by with his name, I'd accept that as well.

blocktrron commented on target/linux/ramips/dts/mt7621_netgear_sercomm_bzv.dtsi:
```diff
@@ -0,0 +1,136 @@
+			pins = "gpio4";
+			bias-pull-up;
+		};
+	};
```
Any reason you have to pull-up this pin? As gpio 4 is used as an LED output (even though the LED is active-low), this seems not necessary to me, as it's not necessary for the other pins.

CHKDSK88 replied:
You're right. Pull-ups are not necessary whet are gpios are push-pull outputs.

blocktrron replied:
The node itself (as well as the referencinf property in the parent node) should now be obsolete then. Should be safe to remove.

CHKDSK88 replied:
> Should be safe to remove.

That's correct. I removed it.

blocktrron commented on target/linux/ramips/dts/mt7621_netgear_sercomm_bzv.dtsi:
```diff
@@ -0,0 +1,136 @@
+		interrupts = <16 0>;
+
+		gpio-controller;
+		interrupt-controller;
```
Enabling interrupt support on an exclusively output-only extender seems strange. Also, i was able to find documentation on interrupt 16 being the interrupt for the i2s bus, while there's no interrupt for i2c. But I'm not an expert with the mt7621.

Anyways, can you remove the interrupt stuff and report back if it works?

CHKDSK88 replied:
I removed it and everything works as expected.

blocktrron reviewed (commented):
I've only just discovered @adrianschmutzler requested me. Sorry for that.

Yohan54 commented:
Just like to note for my AC2400. When compile this branch, selecting R6700v2 and load to router. No issues there at all other than slower speeds using OpenWrt than stock.

It's when revert back to stock AC2400 Netgear Firmware Version 1.2.0.62, it still shows as a R6800 and then have issues where radio's are not enable and cannot get them enable. I then have to download the R6800 Firmware Version 1.2.0.16 to get it to somewhat Netgear stock.

Just to note, yes AC2400 is a clone of R6800 and is a exclusive Best Buy version.

CHKDSK88 commented:
@blocktrron

Is now expander part of dts ok?

CHKDSK88 commented:
Rebased to latest upstream master.

gigul77 commented:
One think which is not working is back to stack firmware... now I've got bricked device. I've compile r6800 and after installing luci a decide to wait for full version available via OpenWRT and sysupgrade to R6800-V1.2.0.62_1.0.1.img ( http://www.downloads.netgear.com/files/GDC/R6800/R6800-V1.2.0.62_1.0.1.zip)...Now it's get up to failsafe but no arp on network so 192.168.1.1 is not working :( Tomorrow I will try serial port installed.

CHKDSK88 commented:
@gigul77
Did You use sysupgrade to back to stock? 

Or maybe nmrpflash?

gigul77 commented:
@CHKDSK88 
I've use "easy" method - install firmware via Luci web interface.

CHKDSK88 commented:
But how you got back to original firmware and brick?

Yohan54 commented:
@CHKDSK88 I had the same issues as @gigul77, only way to fix it was use nmrpflash. I would use sysupgrade -F command and luci as well (basically same thing).
Granted even OpenWrt documentation back stock not clearly defined enough as well.

gigul77 commented:
@CHKDSK88 
I uploaded file via web interface there was warning about mismatch... and now it's brick. It almost go to failsafe method but no answer on 192.168.1.1

Yohan54 commented:
@gigul77 If Netgear router is still bricked, use the nmrpflash, tool works for me every time.
https://github.com/jclehner/nmrpflash > Documentation instructions
https://github.com/jclehner/nmrpflash/releases - Binaries for Linux/Windows

CHKDSK88 commented:
This router have only one way to go back to stock: nmrpflash.

Did You read the commit and PR message?

gigul77 commented:
@CHKDSK88 and @Yohan54 
Thx for help!
No I did not read that. I mean time I solder serial pin and mhhh now it's a real brick :( probably overheated board and now has no life. Tomorrow I will inspect board maybe It will be alive.

CHKDSK88 commented:
https://openwrt.org/inbox/toh/netgear/netgear_r6800 -> I started making wiki page. Serial is described. You need some solder skill to make it.

I will fill it when PR will be merged.

CHKDSK88 commented:
@Yohan54 

>Granted even OpenWrt documentation back stock not clearly defined enough as well.

OpenWrt wiki page have big warning: 
> This page is very old and seems specific to “ancient” devices.
> 
> Consult the device page for more up-to-date and device-specific information.
> 

At this moment this router have description only in commit message.

Yohan54 commented:
@CHKDSK88 @gigul77 I for one wouldn't focus on soldering a serial pin to router board. I would just see if get's to state of "amber" power led. Then, follow the nmrpflash instructions to see if router is responding. I have used the tool every time without any issues when testing after doing a compile of OpenWrt.

gigul77 commented:
@Yohan54 
No amber led any more :(

CHKDSK88 commented:
@gigul77 

https://openwrt.org/_detail/media/netgear/r6800_serial.jpg?id=inbox%3Atoh%3Anetgear%3Anetgear_r6800

Please check if You have every resistor on Your board. It could be configuration pullup/pulldown.

Or maybe You short come connections.

Yohan54 commented:
> @Yohan54
> No amber led any more :(

Any leds at alll, if not, then may have fried the traces when adding the serial jumper and/or solder bridge joints on the serial jumper. By chance, you have serial cable connected when powering on router. I think is only three pins that need to be connected.

blocktrron commented:
Merged to my staging tree - Thanks!

artgru commented:
Hello!
I just installed the r800 openwrt image from this link
http://downloads.openwrt.org/snapshots/targets/ramips/mt7621/openwrt-ramips-mt7621-netgear_r6800-squashfs-factory.img

but I can't access the web gui. How can I solve this?

CHKDSK88 commented:
Snapshot images haven't gui by default. You have to install luci.
```opkg update
opkg install luci
```

artgru commented:
Thanks for the hint! I figured that something is wrong by checking netstat... Now I need to think how to connect the router to the internet as Orange PL Fiber requires additional settings for internet to work..

CHKDSK88 commented:
On polish site "https://eko.one.pl/forum" was discussion about it. You can find solution there.

joelp64 commented:
I have a Netgear R7200.  On netgear's website it uses the same firmware as the R6800, and I opened it up and it looks like the identical board and chips to the R6800, so I figured I would try this firmware.  I installed the snapshot on the Netgear R7200 on 3/31.  That image boots, and let me do a sysupgrade ,and load Luci.  It ran and seemed to work as an access point.  When I tried to configure the router as both a client and access point at the same time it would disable the 2.4ghz radio.  

I tried loading the newest snapshot a couple days ago, and found it would not even boot after the nmrp transfer.  Attached boot log is a boot on the 3/31 snapshot image, a reboot, nmrp flash and subsequent reboot till the system hangs.  

I am more of a hardware guy, so I can't contribute much to fix the problem, but am happy to test and provide feedback.  I added the serial port to the board which is how I got the attached log.   


[boot_nmrp_update_log.txt](https://github.com/openwrt/openwrt/files/4472189/boot_nmrp_update_log.txt)

CHKDSK88 commented:
@joelp64 

Thank You for information. Couple days ago was made huge change in snapshot images and some boards have problem at this moment.

@dengqf6 @981213 
>boot_nmrp_update_loga.txt
Could You look at this mt7621 error?

981213 commented:
@joelp64 @CHKDSK88 
Would you mind to test this patch and see if it fixes the problem?
https://patchwork.ozlabs.org/project/openwrt/patch/20200414064438.1172618-1-gch981213@gmail.com/

joelp64 commented:
@981213 
I would be happy to test it, but I am a hardware guy.  I do not have a build environment set up.  If you could provide me with an .img file I can test it, and post logs, but I don't have the ability to build the files.

joelp64 commented:
Unrelated to the current snapshots not booting, but trying to aid the effort to expand support for these netgear routers, I found this post on another site discussing these routers,  https://www.snbforums.com/threads/netgear-nighthawk-ac2100-smart-wifi-router.49749/#post-442492
The linked post suggests the firmware may identify the specific router via the PCI id in the radio.

From Netgear's site I can confirm that all of these routers link to the same factory firmware.
AC2100
AC2400
AC2600
R6700v2
R6800
R7200 

Only difference in these routers seems to be the number of streams supported, and on some the presence of a USB2.0 port.

CHKDSK88 commented:
@981213 
I sent my device for orphans, so I can't try at my hardware. But I can prepare images for other testers.

@joelp64 

>I would be happy to test it, but I am a hardware guy. I do not have a build environment set up. If you could provide me with an .img file I can test it, and post logs, but I don't have the ability to build the files.

Please unpack zip try image from attachment via nmrpflash.
[openwrt-ramips-mt7621-netgear_r6700-v2-squashfs-factory.img.zip](https://github.com/openwrt/openwrt/files/4477255/openwrt-ramips-mt7621-netgear_r6700-v2-squashfs-factory.img.zip)

>The linked post suggests the firmware may identify the specific router via the PCI id in the radio.

This is interesting. But I'm not sure if in OpenWrt there is any difference in driver.

joelp64 commented:
> Please unpack zip try image from attachment via nmrpflash.
> [openwrt-ramips-mt7621-netgear_r6700-v2-squashfs-factory.img.zip](https://github.com/openwrt/openwrt/files/4477255/openwrt-ramips-mt7621-netgear_r6700-v2-squashfs-factory.img.zip)
The image you sent loaded (nmrp) and booted up fine.  Log attached.
[boot_4_14_test.txt](https://github.com/openwrt/openwrt/files/4477361/boot_4_14_test.txt)

CHKDSK88 commented:
@981213 

Your patch fix the issue.

@joelp64 
Thank You for help. We found third clone of R6700v2. Could You confirm, that every part of hardware (except wifi id) is is the same like R6700v2?

joelp64 commented:
> Thank You for help. We found third clone of R6700v2. Could You confirm, that every part of hardware (except wifi id) is is the same like R6700v2?

I can confirm that everything is identical to internal pictures I have seen for the R6800, except the R7200 does not have the USB 2.0 port.  Photos of the R6800 internal found here: http://koolshare.cn/thread-109522-1-1.html

 I couldn't find any good internal photos of the R6700v2 to compare.  
Reviewing the data here: https://openwrt.org/toh/hwdata/netgear/netgear_r6700_2.0 there is a discrepancy in the number of buttons.  The R7200 has a power switch, a recessed reset button on the back, and two buttons on the top, wifi on/off, and WPS.  For sw purposes it may be that the power and reset switches are not counted as they are not processor inputs.

981213 commented:
@joelp64 @CHKDSK88 
Thanks a lot for your effort!

CHKDSK88 commented:
@joelp64 

R6700v2 have the same buttons as R6800.  So R7200 is R6700v2 clone.

adrianschmutzler commented:
So, just send a PR for it as soon as mt7621 is stable enough for testing again.

CHKDSK88 commented:
Commit is ready. I will wait to @joelp64 answer until tomorrow.

joelp64 commented:
@CHKDSK88 My name is Joel Pinsker.  Is there a way I can share my email with out posting it in an open forum here?  I am trying to avoid the extra spam.

CHKDSK88 commented:
@joelp64 

Commit messages are open posted. So to avoid post e-mail address to open place, I put Your name and link to Your github profile only.

joelp64 commented:
@chkdsk 

Thanks
